File: ios/CookieValidationLogic.swift
import Foundation

enum CookieValidationLogic {
  static func validateStructuredStrings(
    name: String,
    value: String,
    domain: String?,
    path: String?,
    version: String?,
    expires: String?,
    sameSite: String?,
    validate: Bool,
    parseDate: (String) -> Date?
  ) throws {
    guard !name.isEmpty else {
      throw CookieValidationError.invalid("Cookie name must not be empty")
    }

    let fields: [(String, String?)] = [
      ("name", name),
      ("value", value),
      ("domain", domain),
      ("path", path),
      ("version", version),
      ("expires", expires),
      ("sameSite", sameSite),
    ]
    for (field, fieldValue) in fields {
      if let fieldValue, containsASCIIControlCharacter(fieldValue) {
        throw CookieValidationError.invalid(
          "Cookie \(field) must not contain control characters"
        )
      }
    }

    guard !name.contains("=") && !name.contains(";") else {
      throw CookieValidationError.invalid(
        "Cookie name must not contain \"=\" or \";\""
      )
    }
    guard !value.contains(";") else {
      throw CookieValidationError.invalid(
        "Cookie value must not contain \";\" because native stores treat it as an attribute separator"
      )
    }
    guard domain?.contains(";") != true else {
      throw CookieValidationError.invalid("Cookie domain must not contain \";\"")
    }
    guard path?.contains(";") != true else {
      throw CookieValidationError.invalid("Cookie path must not contain \";\"")
    }

    guard validate else { return }

    guard isPortableCookieName(name) else {
      throw CookieValidationError.invalid(
        "Cookie name contains characters unsupported by native stores"
      )
    }
    if let path, !path.isEmpty, !path.hasPrefix("/") {
      throw CookieValidationError.invalid("Cookie path must start with \"/\"")
    }
    if let expires, !expires.isEmpty, parseDate(expires) == nil {
      throw CookieValidationError.invalid(
        "Cookie expires must be a valid ISO 8601 date"
      )
    }
  }

  static func validateRawSetCookieHeader(_ cookie: String) throws {
    guard !cookie.isEmpty else {
      throw CookieValidationError.invalid("Cookie header must not be empty")
    }
    let containsHeaderTerminator = cookie.unicodeScalars.contains { scalar in
      scalar.value == 0 || scalar.value == 10 || scalar.value == 13
    }
    guard !containsHeaderTerminator else {
      throw CookieValidationError.invalid(
        "Cookie header must not contain CR, LF, or NUL"
      )
    }
  }

  private static func containsASCIIControlCharacter(_ value: String) -> Bool {
    value.unicodeScalars.contains { scalar in
      scalar.value <= 0x1F || scalar.value == 0x7F
    }
  }

  private static func isPortableCookieName(_ name: String) -> Bool {
    guard !name.hasPrefix("$") else { return false }

    let punctuation = "!#$%&'*+-.^_`|~"
    return name.unicodeScalars.allSatisfy { scalar in
      switch scalar.value {
      case 0x30...0x39, 0x41...0x5A, 0x61...0x7A:
        return true
      default:
        return scalar.isASCII && punctuation.unicodeScalars.contains(scalar)
      }
    }
  }
}

private enum CookieValidationError: LocalizedError {
  case invalid(String)

  var errorDescription: String? {
    switch self {
    case let .invalid(message): message
    }
  }
}

File: android/src/main/java/com/preeternal/reactnativecookiemanager/CookieValidationLogic.kt
package com.preeternal.reactnativecookiemanager

internal data class StructuredCookieStrings(
  val name: String,
  val value: String,
  val domain: String?,
  val path: String?,
  val version: String?,
  val expires: String?,
  val sameSite: String?
)

internal fun validateStructuredCookieStrings(
  input: StructuredCookieStrings,
  validate: Boolean
) {
  if (input.name.isEmpty()) {
    throw IllegalArgumentException("Cookie name must not be empty")
  }

  listOfNotNull(
    "name" to input.name,
    "value" to input.value,
    input.domain?.let { "domain" to it },
    input.path?.let { "path" to it },
    input.version?.let { "version" to it },
    input.expires?.let { "expires" to it },
    input.sameSite?.let { "sameSite" to it }
  ).forEach { (field, value) ->
    if (value.any(::isAsciiControlCharacter)) {
      throw IllegalArgumentException("Cookie $field must not contain control characters")
    }
  }

  if (input.name.contains('=') || input.name.contains(';')) {
    throw IllegalArgumentException("Cookie name must not contain \"=\" or \";\"")
  }
  if (input.value.contains(';')) {
    throw IllegalArgumentException(
      "Cookie value must not contain \";\" because native stores treat it as an attribute separator"
    )
  }
  if (input.domain?.contains(';') == true) {
    throw IllegalArgumentException("Cookie domain must not contain \";\"")
  }
  if (input.path?.contains(';') == true) {
    throw IllegalArgumentException("Cookie path must not contain \";\"")
  }

  if (!validate) {
    return
  }

  if (!isPortableCookieName(input.name)) {
    throw IllegalArgumentException("Cookie name contains characters unsupported by native stores")
  }
  if (!input.path.isNullOrEmpty() && !input.path.startsWith('/')) {
    throw IllegalArgumentException("Cookie path must start with \"/\"")
  }
  if (!input.expires.isNullOrEmpty() && parseCookieExpires(input.expires) == null) {
    throw IllegalArgumentException("Cookie expires must be a valid ISO 8601 date")
  }
}

internal fun validateRawSetCookieHeader(cookie: String) {
  if (cookie.isEmpty()) {
    throw IllegalArgumentException("Cookie header must not be empty")
  }
  if (cookie.any { it == '\r' || it == '\n' || it == '\u0000' }) {
    throw IllegalArgumentException("Cookie header must not contain CR, LF, or NUL")
  }
}

private fun isAsciiControlCharacter(character: Char): Boolean =
  character.code <= 0x1F || character.code == 0x7F

private fun isPortableCookieName(name: String): Boolean =
  !name.startsWith('$') && name.all { character ->
    character in 'a'..'z' ||
      character in 'A'..'Z' ||
      character in '0'..'9' ||
      character in "!#$%&'*+-.^_`|~"
  }
